packages/superset-ui-translation/src/Translator.ts

```diff
@@ -23,25 +36,62 @@ const DEFAULT_LANGUAGE_PACK = {
 export default class Translator {
   i18n: Jed;
 
+  locale: Locale;
+
   constructor(config: TranslatorConfig = {}) {
     const { languagePack = DEFAULT_LANGUAGE_PACK } = config;
     // eslint-disable-next-line @typescript-eslint/no-unsafe-call
-    this.i18n = new UntypedJed(languagePack) as Jed;
+    this.i18n = new UntyppedJed(languagePack) as Jed;
+    this.locale = this.i18n.options.locale_data.superset[''].lang as Locale;
+  }
+
+  /**
+   * Add additional translations on the fly, used by plugins.
+   */
+  addTranslation(key: string, texts: ReadonlyArray<string>) {
+    const translations = this.i18n.options.locale_data.superset;
+    if (key in translations) {
+      logging.warn(`Duplicate translation key "${key}", will override.`);
+    }
+    translations[key] = texts;
+  }
+
+  /**
+   * Add a series of translations.
+   */
+  addTranslations(translations: Translations) {
+    if (translations && !Array.isArray(translations)) {
+      Object.entries(translations).forEach(([key, vals]) => this.addTranslation(key, vals));
+    } else {
+      logging.warn('Invalid translations');
+    }
+  }
+
+  addLocaleData(data: LocaleData) {
+    // always fallback to English
+    const translations = data?.[this.locale] || data?.en;
+    if (translations) {
+      this.addTranslations(translations);
+    } else {
+      logging.warn('Invalid locale data');
+    }
   }
 
   translate(input: string, ...args: unknown[]): string {
     return this.i18n.translate(input).fetch(...args);
   }
 
-  translateWithNumber(
-    singular: string,
-    plural: string,
-    num: number = 0,
-    ...args: unknown[]
-  ): string {
+  translateWithNumber(key: string, ...args: unknown[]): string {
+    const [plural, num, ...rest] = args;
+    if (typeof plural === 'number') {
+      return this.i18n
+        .translate(key)
+        .ifPlural(plural, key)
+        .fetch(plural, num, ...args);
+    }
```

Comment:
Allow `tn('key', 2)` so we don't have to provide the key twice (`t('key', 'key', 2)`).